metrics/middleware.go

```diff
@@ -2,21 +2,48 @@ package metrics
 
 import (
 	"context"
+	"fmt"
 	"time"
 
 	"github.com/contribsys/faktory/manager"
 	"github.com/contribsys/faktory/util"
 )
 
 func (m *MetricsSubsystem) addMiddleware() {
+	m.Server.Manager().AddMiddleware("push", func(ctx context.Context, next func() error) error {
+		mh := ctx.Value(manager.MiddlewareHelperKey).(manager.Context)
+		tags := m.getTagsFromJob(mh)
+
+		if err := m.StatsDClient().Incr("faktory.jobs.pushed", tags, float64(1)); err != nil {
+			util.Warnf("unable to submit metric: %v", err)
+		}
+
+		return next()
+	})
+
+	m.Server.Manager().AddMiddleware("fetch", func(ctx context.Context, next func() error) error {
+		mh := ctx.Value(manager.MiddlewareHelperKey).(manager.Context)
+		tags := m.getTagsFromJob(mh)
+
+		if err := m.StatsDClient().Incr("faktory.jobs.fetched", tags, float64(1)); err != nil {
+			util.Warnf("unable to submit metric: %v", err)
+		}
+
+		return next()
+	})
+
 	m.Server.Manager().AddMiddleware("ack", func(ctx context.Context, next func() error) error {
 		mh := ctx.Value(manager.MiddlewareHelperKey).(manager.Context)
 		tags := m.getTagsFromJob(mh)
+		tags = append(tags, "status:success")
 
 		if mh.Reservation() != nil {
 			if err := m.StatsDClient().Timing(m.PrefixMetricName("succeeded.time"), time.Duration(time.Since(mh.Reservation().ReservedAt())), tags, 1); err != nil {
 				util.Warnf("unable to submit metric: %v", err)
 			}
+			if err := m.StatsDClient().Timing("faktory.jobs.processed", time.Duration(time.Since(mh.Reservation().ReservedAt())), tags, 1); err != nil {
```

Comment:
Should this one be a `Gauge`?

Reply:
I don't think so. If it were a `Gauge` then we'd only have metrics about how long the most recently completed task ran for. I'd rather have a histogram so that we can see the distribution of runtimes for each task type / queue.